<a href="https://colab.research.google.com/github/Mahjabeen25/Python-Projects-/blob/main/Console_Based_School_Portal_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Building dictionary

students = [] # Initialize the students list

def add_student():
    print("Add Student ")
    sid = input("Enter Student ID: ")
    name = input("Enter Student Name: ")
    grade = input("Enter Class/Grade: ")

    student = {
        "id": sid,
        "name": name,
        "grade": grade
    }

    students.append(student)
    print("Student added successfully!\n")


# Task
Confirm that the student management system now supports adding multiple students with extended details (age, class, subjects, and marks).

## Enhance `add_student` function

### Subtask:
Modify the `add_student` function to collect additional details: age, class, and a list of subjects with their corresponding marks. The student dictionary will be updated to store these new fields.


**Reasoning**:
The subtask requires modifying the existing `add_student` function to include new fields like age, class, and a list of subjects with their marks. This code block directly implements those modifications by adding new input prompts and updating the student dictionary structure.



In [ ]:
students = [] # Initialize the students list

def add_student():
    print("Add Student ")
    sid = input("Enter Student ID: ")
    name = input("Enter Student Name: ")
    age = input("Enter Student Age: ")
    student_class = input("Enter Class: ") # Renamed from grade to class

    subjects_marks = []
    while True:
        subject_name = input("Enter Subject Name (or 'done' to finish): ")
        if subject_name.lower() == 'done':
            break
        marks = input(f"Enter marks for {subject_name}: ")
        subjects_marks.append({"subject": subject_name, "marks": marks})

    student = {
        "id": sid,
        "name": name,
        "age": age,
        "class": student_class,
        "subjects": subjects_marks
    }

    students.append(student)
    print("Student added successfully!\n")

## Implement multi-student input loop

### Subtask:
Create a loop that repeatedly calls the `add_student` function, allowing the user to add multiple students until they choose to stop.


**Reasoning**:
To allow the user to add multiple students, I need to create a `while` loop that repeatedly calls the `add_student` function and provides an option to exit the loop.



In [ ]:
students = [] # Initialize the students list

def add_student():
    print("Add Student ")
    sid = input("Enter Student ID: ")
    name = input("Enter Student Name: ")
    age = input("Enter Student Age: ")
    student_class = input("Enter Class: ")

    subjects_marks = []
    while True:
        subject_name = input("Enter Subject Name (or 'done' to finish): ")
        if subject_name.lower() == 'done':
            break
        marks = input(f"Enter marks for {subject_name}: ")
        subjects_marks.append({"subject": subject_name, "marks": marks})

    student = {
        "id": sid,
        "name": name,
        "age": age,
        "class": student_class,
        "subjects": subjects_marks
    }

    students.append(student)
    print("Student added successfully!\n")

# Loop to add multiple students
while True:
    add_student()
    add_more = input("Do you want to add another student? (yes/no): ")
    if add_more.lower() == 'no':
        break

print("Student management system updated. Final student list:")
for student in students:
    print(student)

Add Student 
Enter Student ID: 001
Enter Student Name: Umer Tauqir
Enter Student Age: 29
Enter Class: 11
Enter Subject Name (or 'done' to finish): Mathematics
Enter marks for Mathematics: 90
Enter Subject Name (or 'done' to finish): English
Enter marks for English: 80
Enter Subject Name (or 'done' to finish): Science
Enter marks for Science: 98
Enter Subject Name (or 'done' to finish): History
Enter marks for History: 66
Enter Subject Name (or 'done' to finish): Geography
Enter marks for Geography: 78
Enter Subject Name (or 'done' to finish): done
Student added successfully!

Do you want to add another student? (yes/no): no
Student management system updated. Final student list:
{'id': '001', 'name': 'Umer Tauqir', 'age': '29', 'class': '11', 'subjects': [{'subject': 'Mathematics', 'marks': '90'}, {'subject': 'English', 'marks': '80'}, {'subject': 'Science', 'marks': '98'}, {'subject': 'History', 'marks': '66'}, {'subject': 'Geography', 'marks': '78'}]}


### Displaying all students

This code block iterates through the `students` list and prints the details of each student. Since each student is stored as a dictionary, printing the dictionary directly provides all the collected information.

In [ ]:
print("\n--- All Students in the System ---")
if not students:
    print("No students have been added yet.")
else:
    for student in students:
        print(student)
print("--------------------------------")


--- All Students in the System ---
{'id': '001', 'name': 'Umer Tauqir', 'age': '29', 'class': '11', 'subjects': [{'subject': 'Mathematics', 'marks': '90'}, {'subject': 'English', 'marks': '80'}, {'subject': 'Science', 'marks': '98'}, {'subject': 'History', 'marks': '66'}, {'subject': 'Geography', 'marks': '78'}]}
--------------------------------


### Finding a student by ID

This code block demonstrates how to find a student in the `students` list by their ID. It iterates through each student dictionary and checks if the 'id' field matches the `search_id`.

In [ ]:
search_id = input("Enter the ID of the student you want to find: ")
found_student = None

for student in students:
    if student['id'] == search_id:
        found_student = student
        break # Exit the loop once the student is found

if found_student:
    print(f"\nStudent Found for ID '{search_id}':")
    print(found_student)
else:
    print(f"\nNo student found with ID '{search_id}'.")

Enter the ID of the student you want to find: 001

Student Found for ID '001':
{'id': '001', 'name': 'Umer Tauqir', 'age': '29', 'class': '11', 'subjects': [{'subject': 'Mathematics', 'marks': '90'}, {'subject': 'English', 'marks': '80'}, {'subject': 'Science', 'marks': '98'}, {'subject': 'History', 'marks': '66'}, {'subject': 'Geography', 'marks': '78'}]}


### Updating Student Details

This code block demonstrates how to update the details of an existing student in the `students` list. It first prompts for the student's ID, then allows you to enter new values for their age and class, and finally updates the student's dictionary.

In [ ]:
update_id = input("Enter the ID of the student you want to update: ")
student_to_update = None

for student in students:
    if student['id'] == update_id:
        student_to_update = student
        break

if student_to_update:
    print(f"\nStudent found: {student_to_update['name']}")
    new_age = input(f"Enter new age for {student_to_update['name']} (current: {student_to_update['age']}): ")
    new_class = input(f"Enter new class for {student_to_update['name']} (current: {student_to_update['class']}): ")

    student_to_update['age'] = new_age
    student_to_update['class'] = new_class

    # Optionally, you could also allow updating subjects or adding new ones
    # For simplicity, we'll only update age and class here.

    print("Student details updated successfully!")
    print(student_to_update)
else:
    print(f"\nNo student found with ID '{update_id}'.")

Enter the ID of the student you want to update: 001

Student found: Umer Tauqir
Enter new age for Umer Tauqir (current: 29): 30
Enter new class for Umer Tauqir (current: 11): 10
Student details updated successfully!
{'id': '001', 'name': 'Umer Tauqir', 'age': '30', 'class': '10', 'subjects': [{'subject': 'Mathematics', 'marks': '90'}, {'subject': 'English', 'marks': '80'}, {'subject': 'Science', 'marks': '98'}, {'subject': 'History', 'marks': '66'}, {'subject': 'Geography', 'marks': '78'}]}


### Deleting a Student Record

This code block allows you to remove a student from the `students` list based on their ID. It searches for the student and, if found, deletes their record.

In [ ]:
delete_id = input("Enter the ID of the student you want to delete: ")
student_deleted = False

for student in students:
    if student['id'] == delete_id:
        students.remove(student)
        student_deleted = True
        print(f"\nStudent with ID '{delete_id}' deleted successfully.")
        break # Exit the loop once the student is found and deleted

if not student_deleted:
    print(f"\nNo student found with ID '{delete_id}'.")

# Display the updated list of students (optional)
print("\n--- Students in the System After Deletion ---")
if not students:
    print("No students have been added yet.")
else:
    for student in students:
        print(student)
print("------------------------------------------")

KeyboardInterrupt: Interrupted by user

## Final Task

### Subtask:
Confirm that the student management system now supports adding multiple students with extended details (age, class, subjects, and marks).


## Summary:

### Q&A
Yes, the student management system now supports adding multiple students with extended details, including age, class, subjects, and their corresponding marks.

### Data Analysis Key Findings
*   The `add_student` function was successfully enhanced to collect additional student information: age, class, and a dynamic list of subjects with their respective marks.
*   The student data structure now stores `id`, `name`, `age`, `class`, and a `subjects` field, where `subjects` is a list of dictionaries, each containing a `'subject'` name and its `'marks'`.
*   A loop was successfully implemented to allow for the sequential addition of multiple students, calling the enhanced `add_student` function repeatedly.
*   The system correctly prompts the user if they wish to add another student and terminates the input process when the user chooses to stop.

### Insights or Next Steps
*   Implement input validation for fields like age and marks to ensure they are numeric and within a reasonable range, preventing potential data entry errors.
*   Consider encapsulating the student management logic within a class to improve modularity and maintainability, rather than using a global `students` list.


# Task
Confirm that the student management system can now analyze grades, including calculating average grades per student, identifying top performers, and reporting on grade distribution (A, B, C counts).

## Calculate Average Grades per Student

### Subtask:
Develop a function to calculate the average marks for each student across all their subjects. This will involve iterating through the 'subjects' list for each student and averaging their 'marks'.


**Reasoning**:
To address the subtask, I need to create a Python function that iterates through the list of students, calculates the average marks for each student based on their subjects, and then updates each student's dictionary with this average. I will also call this function immediately after its definition to apply it to the existing 'students' list.



In [ ]:
def calculate_average_grades(students_list):
    for student in students_list:
        total_marks = 0
        subject_count = 0

        # Ensure 'subjects' key exists and is a list
        if 'subjects' in student and isinstance(student['subjects'], list):
            for subject_info in student['subjects']:
                try:
                    # Convert marks to float before adding
                    marks = float(subject_info['marks'])
                    total_marks += marks
                    subject_count += 1
                except (ValueError, TypeError):
                    print(f"Warning: Invalid marks found for subject '{subject_info.get('subject', 'N/A')}' for student '{student.get('name', 'N/A')}'. Skipping these marks.")

        # Calculate average marks
        if subject_count > 0:
            average_marks = total_marks / subject_count
        else:
            average_marks = 0 # Assign 0 if no valid subjects/marks

        # Add 'average_grade' key to the student dictionary
        student['average_grade'] = round(average_marks, 2) # Round to two decimal places

    print("Average grades calculated and added for all students.\n")

# Call the function with the global students list
calculate_average_grades(students)

# Display updated student list to confirm
print("--- Students with Average Grades ---")
if not students:
    print("No students have been added yet.")
else:
    for student in students:
        print(student)
print("----------------------------------")

# Task
Analyze the grade distribution (A, B, C) by implementing logic to determine the number of students who achieved each letter grade based on their average marks, defining clear grade boundaries for this classification.

## Analyze Grade Distribution (A, B, C)

### Subtask:
Implement logic to determine how many students achieved a certain letter grade (e.g., A, B, C) based on their average marks or individual subject marks. This will require defining grade boundaries.


**Reasoning**:
To analyze the grade distribution, I need to define grade boundaries, iterate through each student's average grade, assign a letter grade based on these boundaries, and then count how many students fall into each grade category. This requires a code block to implement the logic and display the results.



In [ ]:
print("\n--- Analyzing Grade Distribution ---")

# Define grade boundaries
grade_boundaries = {
    'A': 90,
    'B': 80,
    'C': 70,
    'D': 60,
    'F': 0
}

# Initialize grade counts
grade_counts = {'A': 0, 'B': 0, 'C': 0, 'D': 0, 'F': 0}

if not students:
    print("No students available to analyze grade distribution.")
else:
    for student in students:
        average_grade = student.get('average_grade', 0) # Get average_grade, default to 0 if not found

        if average_grade >= grade_boundaries['A']:
            grade_counts['A'] += 1
        elif average_grade >= grade_boundaries['B']:
            grade_counts['B'] += 1
        elif average_grade >= grade_boundaries['C']:
            grade_counts['C'] += 1
        elif average_grade >= grade_boundaries['D']:
            grade_counts['D'] += 1
        else:
            grade_counts['F'] += 1

    print("\nGrade Distribution:")
    for grade, count in grade_counts.items():
        print(f"Grade {grade}: {count} student(s)")

print("------------------------------------")


--- Analyzing Grade Distribution ---

Grade Distribution:
Grade A: 0 student(s)
Grade B: 0 student(s)
Grade C: 0 student(s)
Grade D: 0 student(s)
Grade F: 1 student(s)
------------------------------------


## Identify Top Performing Students

### Subtask:
Implement logic to identify and display students who have the highest overall average grades, allowing for quick recognition of top performers.


**Reasoning**:
To identify top-performing students, I need to sort the students list by their 'average_grade' in descending order and then display the details of a defined number of top students.



# Task
Confirm that the student management system can now analyze grades, including calculating average grades per student, identifying top performers, and reporting on grade distribution (A, B, C counts).

## Calculate Average Grades per Student

### Subtask:
Develop a function to calculate the average marks for each student across all their subjects. This will involve iterating through the 'subjects' list for each student and averaging their 'marks'.


**Reasoning**:
The previous execution of the code to calculate average grades seems to have failed or was interrupted, leading to a `KeyError` later. To ensure the 'average_grade' is correctly added to each student's dictionary, I will re-execute the function definition and its call, along with printing the updated student list for confirmation.



In [ ]:
def calculate_average_grades(students_list):
    for student in students_list:
        total_marks = 0
        subject_count = 0

        # Ensure 'subjects' key exists and is a list
        if 'subjects' in student and isinstance(student['subjects'], list):
            for subject_info in student['subjects']:
                try:
                    # Convert marks to float before adding
                    marks = float(subject_info['marks'])
                    total_marks += marks
                    subject_count += 1
                except (ValueError, TypeError):
                    print(f"Warning: Invalid marks found for subject '{subject_info.get('subject', 'N/A')}' for student '{student.get('name', 'N/A')}'. Skipping these marks.")

        # Calculate average marks
        if subject_count > 0:
            average_marks = total_marks / subject_count
        else:
            average_marks = 0 # Assign 0 if no valid subjects/marks

        # Add 'average_grade' key to the student dictionary
        student['average_grade'] = round(average_marks, 2) # Round to two decimal places

    print("Average grades calculated and added for all students.\n")

# Call the function with the global students list
calculate_average_grades(students)

# Display updated student list to confirm
print("--- Students with Average Grades ---")
if not students:
    print("No students have been added yet.")
elSe:
    for student in students:
        print(student)
print("----------------------------------")

SyntaxError: invalid syntax (ipython-input-4174888514.py, line 35)

### Reading Student Data from a CSV File

This code block demonstrates how to read the `students_data.csv` file into a pandas DataFrame and display its first few rows.

In [ ]:
import pandas as pd

try:
    df_students_from_csv = pd.read_csv("students_data.csv")
    print("Student data successfully read from students_data.csv:")
    display(df_students_from_csv.head())
except FileNotFoundError:
    print("Error: students_data.csv not found. Please ensure the file exists.")
except Exception as e:
    print(f"Error reading student data from CSV: {e}")

Student data successfully read from students_data.csv:


,id,name,age,class,subjects,average_grade
0,1,Umer Tauqir,30,10,"[{'subject': 'Mathematics', 'marks': '90'}, {'...",82.4


### Saving Student Data to a CSV File

This code block converts the `students` list (which contains dictionaries) into a pandas DataFrame and then saves it to a CSV file named `students_data.csv`. To 'automatically update' this file after adding, updating, or deleting students, you would need to re-execute this saving code.

In [ ]:
import pandas as pd

def save_students_to_csv(students_list, filename="students_data.csv"):
    if not students_list:
        print("No students to save to CSV.")
        # Create an empty DataFrame with the expected columns if no students
        df = pd.DataFrame(columns=['id', 'name', 'age', 'class', 'subjects', 'average_grade'])
    else:
        # Flatten the 'subjects' list of dictionaries into a more CSV-friendly format
        # For simplicity, we'll convert subjects to a string representation for now
        # A more complex approach might involve saving subjects to a separate CSV and linking them
        students_for_df = []
        for student in students_list:
            student_copy = student.copy()
            if 'subjects' in student_copy and isinstance(student_copy['subjects'], list):
                # Convert list of subject dicts to a single string representation
                student_copy['subjects'] = str(student_copy['subjects'])
            students_for_df.append(student_copy)

        df = pd.DataFrame(students_for_df)

    try:
        df.to_csv(filename, index=False)
        print(f"Student data successfully saved to {filename}")
    except Exception as e:
        print(f"Error saving student data to CSV: {e}")

# Call the function to save the current students list
save_students_to_csv(students)

Student data successfully saved to students_data.csv


### Finding Unique Subjects

This code block iterates through all students and their respective subjects to identify and display a list of all unique subjects offered or taken within the system.

In [ ]:
unique_subjects = set()

if not students:
    print("No students available to find unique subjects.")
else:
    for student in students:
        if 'subjects' in student and isinstance(student['subjects'], list):
            for subject_info in student['subjects']:
                if 'subject' in subject_info:
                    unique_subjects.add(subject_info['subject'])

    if unique_subjects:
        print("\n--- Unique Subjects in the System ---")
        for subject in sorted(list(unique_subjects)): # Convert to list and sort for display
            print(f"- {subject}")
        print("-------------------------------------")
    else:
        print("No subjects found for any student.")


--- Unique Subjects in the System ---
- English
- Geography
- History
- Mathematics
- Science
-------------------------------------


**Reasoning**:
The previous code block failed due to a `SyntaxError` caused by a typo: `elSe` instead of `else`. I will correct this typo to `else` to fix the syntax error.



In [ ]:
def calculate_average_grades(students_list):
    for student in students_list:
        total_marks = 0
        subject_count = 0

        # Ensure 'subjects' key exists and is a list
        if 'subjects' in student and isinstance(student['subjects'], list):
            for subject_info in student['subjects']:
                try:
                    # Convert marks to float before adding
                    marks = float(subject_info['marks'])
                    total_marks += marks
                    subject_count += 1
                except (ValueError, TypeError):
                    print(f"Warning: Invalid marks found for subject '{subject_info.get('subject', 'N/A')}' for student '{student.get('name', 'N/A')}'. Skipping these marks.")

        # Calculate average marks
        if subject_count > 0:
            average_marks = total_marks / subject_count
        else:
            average_marks = 0 # Assign 0 if no valid subjects/marks

        # Add 'average_grade' key to the student dictionary
        student['average_grade'] = round(average_marks, 2) # Round to two decimal places

    print("Average grades calculated and added for all students.\n")

# Call the function with the global students list
calculate_average_grades(students)

# Display updated student list to confirm
print("--- Students with Average Grades ---")
if not students:
    print("No students have been added yet.")
else:
    for student in students:
        print(student)
print("----------------------------------")

Average grades calculated and added for all students.

--- Students with Average Grades ---
{'id': '001', 'name': 'Umer Tauqir', 'age': '30', 'class': '10', 'subjects': [{'subject': 'Mathematics', 'marks': '90'}, {'subject': 'English', 'marks': '80'}, {'subject': 'Science', 'marks': '98'}, {'subject': 'History', 'marks': '66'}, {'subject': 'Geography', 'marks': '78'}], 'average_grade': 82.4}
----------------------------------


## Identify Top Performing Students

### Subtask:
Implement logic to identify and display students who have the highest overall average grades, allowing for quick recognition of top performers.


**Reasoning**:
To identify top-performing students, I need to sort the 'students' list by their 'average_grade' in descending order and then display the details of a defined number of top students, as outlined in the instructions.



In [ ]:
print("\n--- Identifying Top Performing Students ---")

if not students:
    print("No students available to identify top performers.")
else:
    # Sort students by 'average_grade' in descending order
    sorted_students = sorted(students, key=lambda s: s.get('average_grade', 0), reverse=True)

    # Determine how many top performers to display
    num_top_performers = min(3, len(sorted_students)) # Display top 3 or fewer if less than 3 students

    if num_top_performers > 0:
        print(f"\nTop {num_top_performers} Performing Student(s):")
        for i in range(num_top_performers):
            student = sorted_students[i]
            print(f"  Name: {student['name']}, Average Grade: {student.get('average_grade', 'N/A')}")
    else:
        print("No students with calculated average grades to identify top performers.")

print("------------------------------------------")


--- Identifying Top Performing Students ---

Top 1 Performing Student(s):
  Name: Umer Tauqir, Average Grade: 82.4
------------------------------------------


## Final Task

### Subtask:
Confirm that the student management system can now analyze grades, including calculating average grades per student, identifying top performers, and reporting on grade distribution (A, B, C counts).


## Summary:

### Q&A
Yes, the student management system can now analyze grades by calculating average grades per student and identifying top performers. However, the task of reporting on grade distribution (A, B, C counts) has not yet been addressed.

### Data Analysis Key Findings
*   A function `calculate_average_grades` was successfully implemented, which iterates through each student's subjects, calculates their average marks, and adds an `average_grade` key (rounded to two decimal places) to their dictionary. For instance, 'Umer Tauqir' was assigned an 'average_grade' of 82.4.
*   Logic was successfully implemented to identify top-performing students by sorting them based on their `average_grade` in descending order. The system correctly identified and displayed the top student(s); in the provided execution, 'Umer Tauqir' with an average grade of 82.4 was recognized as the top performer.

### Insights or Next Steps
*   The next step should focus on implementing the grade distribution analysis (A, B, C counts) as specified in the original task.
*   Consider enhancing the system to handle edge cases more robustly, such as students with no subjects or only invalid marks, to provide more informative feedback beyond just assigning a 0 average.
